### ISSUE TITLE
Can't end BackgroundTask on iOS when launch other application.

### ISSUE THREAD

**Issue by wrl96** [flutter/flutter#41095]
## Details
Xcode outputs "Can't end BackgroundTask: no background task exists with identifier 1 (0x1), or it may have already been ended. Break in UIApplicationEndBackgroundTaskError() to debug." when launch other apps.

## Code
```
import 'package:flutter/material.dart';
import 'package:url_launcher/url_launcher.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage();
  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('UIApplicationEndBackgroundTaskError'),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: () async => await launch('sms:1111111'), // The error will be occured after pressed.
        tooltip: 'Increment',
        child: Icon(Icons.add),
      ),
    );
  }
}
```

## Logs

Xcode outputs:
```
2019-09-23 19:56:29.683983+0800 Runner[35877:35419171] flutter: Observatory listening on http://127.0.0.1:54688/jUqXY95Znm4=/
2019-09-23 19:56:29.891716+0800 Runner[35877:35419154] flutter: Overflow on channel: flutter/lifecycle.  Messages on this channel are being discarded in FIFO fashion.  The engine may not be running or you need to adjust the buffer size if of the channel.
2019-09-23 19:56:36.772622+0800 Runner[35877:35418906] Can't end BackgroundTask: no background task exists with identifier 1 (0x1), or it may have already been ended. Break in UIApplicationEndBackgroundTaskError() to debug.
```

flutter doctor -v:
```
[✓] Flutter (Channel master, v1.10.6-pre.29, on Mac OS X 10.14.5 18F132, locale zh-Hans-CN)
    • Flutter version 1.10.6-pre.29 at /Users/*/Documents/flutter
    • Framework revision 919cc97f6a (3 days ago), 2019-09-20 14:57:05 -0700
    • Engine revision 99092a0436
    • Dart version 2.6.0 (build 2.6.0-dev.0.0 7c1821c4aa)

 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /Users/*/Library/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-29, build-tools 29.0.2
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.0)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.0, Build version 11A420a
    • CocoaPods version 1.7.5

[✓] Android Studio (version 3.5)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 39.0.3
    • Dart plugin version 191.8423
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)

[✓] Connected device (2 available)
    • iPhone       • **** • ios • iOS 13.0
    • iPhone 11 Pro Max • 4D1EE296-9A86-4315-B910-7C2D1EE25493     • ios • com.apple.CoreSimulator.SimRuntime.iOS-13-0 (simulator)

• No issues found!
```

**Comment by BondarenkoStas** [comment]
@wrl96 It may be an OS-level bug causing the warning as it said in https://forums.developer.apple.com/thread/121990. Is it so that you have only this output, without additional crushes?

**Comment by wrl96** [comment]
@BondarenkoStas I have only this output, without crashes.

**Comment by Tabarek** [comment]
I had the same issue

**Comment by jonataslaw** [comment]
I had the same issue on linux

**Comment by billy124** [comment]
Same issue

**Comment by genert** [comment]
Same problem occurs here. 

```
[✓] Flutter (Channel master, v1.10.14-pre.12, on Mac OS X 10.14.6 18G95, locale en-GB)
    • Flutter version 1.10.14-pre.12 at /Users/genertorg/flutter
    • Framework revision cdc2d9901d (11 hours ago), 2019-10-07 16:43:04 -0700
    • Engine revision 1d62160fdb
    • Dart version 2.6.0 (build 2.6.0-dev.1.0 d6c6d12ebf)


[!] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
    • Android SDK at /Users/genertorg/Library/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-28, build-tools 28.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)
    ! Some Android licenses not accepted.  To resolve this, run: flutter doctor --android-licenses

[✓] Xcode - develop for iOS and macOS (Xcode 11.0)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.0, Build version 11A420a
    • CocoaPods version 1.7.5

[✓] Android Studio (version 3.5)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 40.0.2
    • Dart plugin version 191.8423
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)

[✓] IntelliJ IDEA Ultimate Edition (version 2019.2.2)
    • IntelliJ at /Applications/IntelliJ IDEA.app
    • Flutter plugin version 39.0.5
    • Dart plugin version 192.6262.58

[✓] VS Code (version 1.38.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.5.1

[✓] Connected device (1 available)
    • iPhone • 9581bfd8064d6254aaa4c9bccbf2b8055a9a1c83 • ios • iOS 13.1.2
```

**Comment by genert** [comment]
Could this be connected with new lifecycle paradigm? - https://github.com/firebase/firebase-ios-sdk/issues/3521

**Comment by genert** [comment]
Have you encountered such error in iOS 13 for Flutter? - @chinmaygarde @dnfield

**Comment by dnfield** [comment]
@jmagman or @gaaclarke have you noticed this?

Does this cause any other issues?  Does the app run properly.

**Comment by gaaclarke** [comment]
I tried to verify on iOS 11 and didn't see the error.  We are thinking this is iOS 13 then?

The only background task in the engine is here: https://github.com/flutter/engine/blob/5b3182294f6369bbb2b850d43dbb6710c64ca5b2/shell/platform/darwin/ios/framework/Source/FlutterPluginAppLifeCycleDelegate.mm#L143

It only runs in debug mode.  According to this stack overflow discussion: https://stackoverflow.com/questions/23882495/cant-endbackgroundtask-no-background-task-exists-with-identifier-or-it-may-ha we need to set the background task identifier to `UIBackgroundTaskInvalid`.

**Comment by gaaclarke** [comment]
I was able to verify the bug on iOS 13.  If I comment out all instances of our background tasks in FlutterPluginAppLifeCycleDelegate the error still is printed out.  We have no other background tasks in our code.  I've put a breakpoint in UIApplicationEndBackgroundTaskError but when that gets triggered there is no meaningful information in the backtrace:
```
Thread 1 Queue : com.apple.main-thread (serial)
#0	0x000000019247491c in UIApplicationEndBackgroundTaskError ()
#1	0x0000000192474c04 in -[UIApplication _endBackgroundTask:] ()
#2	0x0000000103012c04 in _dispatch_client_callout ()
#3	0x0000000103016028 in _dispatch_block_invoke_direct ()
#4	0x000000019352b540 in __FBSSERIALQUEUE_IS_CALLING_OUT_TO_A_BLOCK__ ()
#5	0x000000019352b20c in -[FBSSerialQueue _queue_performNextIfPossible] ()
#6	0x000000019352b734 in -[FBSSerialQueue _performNextFromRunLoopSource] ()
#7	0x000000018e3c67e0 in __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__ ()
#8	0x000000018e3c6738 in __CFRunLoopDoSource0 ()
#9	0x000000018e3c5ed0 in __CFRunLoopDoSources0 ()
#10	0x000000018e3c101c in __CFRunLoopRun ()
#11	0x000000018e3c08bc in CFRunLoopRunSpecific ()
#12	0x000000019822c328 in GSEventRunModal ()
#13	0x00000001924566d4 in UIApplicationMain ()
#14	0x0000000102f48d78 in main at /Users/aaclarke/dev/flutter-examples/myapp/ios/Runner/AppDelegate.swift:5
#15	0x000000018e24b460 in start ()
```

I put breakpoints in `-[UIApplication endBackgroundTask:]` and `-[UIApplication beginBackgroundTaskWithName:expirationHandler:]` but they are never triggered.  I daresay this may be an iOS bug.  Luckily it doesn't seem to have any negative consequences other than a log message.

**Comment by genert** [comment]
Seems to be iOS 13 only. Did not appear before that upgrade, and other observations seem to confirm that it is iOS 13 only issue.

@dnfield App works perfectly fine. However, when user puts app to background the error will crash all other background activities making for example push notifications (APNs) not work. When app is brought back to foreground, it works fine.

**Comment by genert** [comment]
@gaaclarke Yes, seems to be caused by iOS directly - https://forums.developer.apple.com/thread/121990

Commenting on my observation that push notifications to not work, it seems that others are struggling with that as well, not Flutter specific fortunately - https://github.com/invertase/react-native-firebase/issues/2657

Have to investigate.

**Comment by phranck** [comment]
> Could this be connected with new lifecycle paradigm? - https://stackoverflow.com/questions/56508764/app-delegate-methods-arent-being-called-in-ios-13/56508769#56508769

@genert This is very interesting. Today morning for no reason I discovered, that (suddenly) all these "old fashioned" life cycle app delegate calls no more called in our Flutter app. So I tested the new ones (we still don't use an Scene stuff!) and: Bingo! They work.

Unfortunately I absolutely have no idea neither why they work nor why the old ones doesn't work anymore. 🤔

**Comment by wrl96** [comment]
> > Could this be connected with new lifecycle paradigm? - https://stackoverflow.com/questions/56508764/app-delegate-methods-arent-being-called-in-ios-13/56508769#56508769
> 
> @genert This is very interesting. Today morning for no reason I discovered, that (suddenly) all these "old fashioned" life cycle app delegate calls aren't called anymore in our Flutter app. So I tested the new ones (we still don't use any Scene stuff!) and: Bingo! They work.
> 
> Unfortunately I absolutely have no idea neither why they work nor why the old ones doesn't work anymore. 🤔

This error still in the latest stable version. 1.12.13+hotfix.5

**Comment by rodriigovieira** [comment]
This seems to be already solved by Apple on iOS 13.4 beta 1, as mentioned here: https://github.com/firebase/firebase-ios-sdk/issues/4554#issuecomment-588430305.

Should we close this?

**Comment by jmagman** [comment]
Thanks, @rodriigovieira.
Closing.  If anyone is seeing this on >iOS 13.5 please let us know and we can reopoen.
